bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/AbstractExpression.java

```diff
@@ -0,0 +1,278 @@
+        }
+        return startDate;
+    }
+
+    @Override
+    public void setStartDate(Date startDate) throws IllegalArgumentException, ParseException {
+        if (startDate == null) {
+            throw new IllegalArgumentException("The start date of the rule must not be null");
+        }
+        this.startDate = startDate;
+        parseExpression(expression);
+    }
+
+    @Override
+    public TimeZone getTimeZone() {
+        if (timeZone == null) {
+            timeZone = TimeZone.getDefault();
+        }
+
+        return timeZone;
+    }
+
+    @Override
+    public void setTimeZone(TimeZone timeZone) throws IllegalArgumentException, ParseException {
+        if (timeZone == null) {
+            throw new IllegalArgumentException("The time zone must not be null");
+        }
+        this.timeZone = timeZone;
+        parseExpression(expression);
+    }
+
+    @Override
+    public String getExpression() {
+        return expression;
+    }
+
+    @Override
+    public void setExpression(String expression) throws ParseException {
+        this.expression = expression;
+        parseExpression(expression);
+    }
+
+    @Override
+    public String toString() {
+        return expression;
+    }
+
+    /**
+     * Parse the given expression
+     *
+     * @param expression - the expression to parse
+     * @throws ParseException - when the expression cannot be successfully be parsed
+     * @throws IllegalArgumentException - when expression parts conflict with each other
+     */
+    public void parseExpression(String expression) throws ParseException, IllegalArgumentException {
+
+        StringTokenizer expressionTokenizer = new StringTokenizer(expression, delimiters, false);
+        int position = 0;
+
+        setExpressionParts(new ArrayList<E>());
+        setCandidates(new ArrayList<Date>());
+
+        while (expressionTokenizer.hasMoreTokens()) {
+            String token = expressionTokenizer.nextToken().trim();
+            position++;
+            getExpressionParts().add(parseToken(token, position));
+        }
+
+        Collections.sort(getExpressionParts());
+
+        validateExpression();
+
+        if (startDate == null) {
+            setStartDate(Calendar.getInstance().getTime());
+        }
+
+        applyExpressionParts();
+
+        synchronized (this) {
+            continueSearch = true;
+            while (getCandidates().size() < minimumCandidates && continueSearch) {
+                populateWithSeeds();
+                getCandidates().clear();
+                applyExpressionParts();
+            }
+            continueSearch = false;
+        }
+
+        for (Date aDate : getCandidates()) {
+            logger.trace("Final candidate {} is {}", getCandidates().indexOf(aDate), aDate);
+        }
+    }
+
+    abstract protected void validateExpression() throws IllegalArgumentException;
+
+    protected void applyExpressionParts() {
+        Collections.sort(getExpressionParts());
```

Comment:
Above this is done in front of the validate call, so if it has been sorted before the validation, couldn't we use the expression part as it is instead of calling sort again...